Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp

```diff
@@ -5668,9 +5668,25 @@ auto OMGIRGenerator::addCallRef(const TypeDefinition& originalSignature, Argumen
         m_currentBlock->appendNew<MemoryValue>(m_proc, Load, pointerType(), origin(), callee,
         safeCast<int32_t>(WebAssemblyFunctionBase::offsetOfEntrypointLoadLocation())));
 
-    Value* calleeCallee = m_currentBlock->appendNew<MemoryValue>(m_proc, Load, pointerType(), origin(),
-        m_currentBlock->appendNew<MemoryValue>(m_proc, Load, pointerType(), origin(), callee,
-        safeCast<int32_t>(WebAssemblyFunctionBase::offsetOfBoxedWasmCalleeLoadLocation())));
+    Value* calleeLoadLocation = m_currentBlock->appendNew<MemoryValue>(m_proc, Load, pointerType(), origin(), callee, safeCast<int32_t>(WebAssemblyFunctionBase::offsetOfBoxedWasmCalleeLoadLocation()));
+
+    BasicBlock* nonzero = m_proc.addBlock();
+    BasicBlock* continuation = m_proc.addBlock();
+
+    UpsilonValue* zeroUpsilon = m_currentBlock->appendNew<UpsilonValue>(m_proc, origin(), constant(pointerType(), 0));
+    m_currentBlock->appendNewControlValue(m_proc, B3::Branch, origin(), calleeLoadLocation, FrequentedBlock(nonzero), FrequentedBlock(continuation, FrequencyClass::Rare));
+
+    Value* loadedCallee = nonzero->appendNew<MemoryValue>(m_proc, Load, pointerType(), origin(), calleeLoadLocation);
+    UpsilonValue* nonzeroUpsilon = nonzero->appendNew<UpsilonValue>(m_proc, origin(), loadedCallee);
+    nonzero->appendNewControlValue(m_proc, Jump, origin(), continuation);
+
+    continuation->addPredecessor(m_currentBlock);
+    continuation->addPredecessor(nonzero);
+
+    m_currentBlock = continuation;
+    Value* calleeCallee = m_currentBlock->appendNew<Value>(m_proc, Phi, pointerType(), origin());
+    nonzeroUpsilon->setPhi(calleeCallee);
+    zeroUpsilon->setPhi(calleeCallee);
```

Comment:
Ditto. I wonder if we can put a thunk instead of zero, so that this check is not necessary.